### PR TITLE
feat: Implements `mongodbatlas_flex_cluster` plural data source

### PR DESCRIPTION
**Pull request by cveticm**
## Description

Implements `mongodbatlas_flex_cluster` plural data source and testing.

Link to any related issue(s): [CLOUDP-278906](https://jira.mongodb.org/browse/CLOUDP-278906)

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [x] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR. A migration guide must be created or updated if the new feature will go in a major version.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR. A migration guide must be created or updated.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contributing guides](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/contributing/README.md)
- [x] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [ ] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code
- [ ] If changes include deprecations or removals I have added appropriate changelog entries.
- [ ] If changes include removal or addition of 3rd party GitHub actions, I updated our internal document. Reach out to the APIx Integration slack channel to get access to the internal document.

### CODE REVIEW

File: internal/service/flexcluster/resource_test.go [hunk 1]
```diff
@@ -13,19 +13,20 @@ import (
...
+	resourceType         = "mongodbatlas_flex_cluster"
+	resourceName         = "mongodbatlas_flex_cluster.test"
+	dataSourceName       = "data.mongodbatlas_flex_cluster.test"
+	dataSourcePluralName = "data.mongodbatlas_flex_clusters.test"
 )
 
 func TestAccFlexClusterRS_basic(t *testing.T) {
 	tc := basicTestCase(t)
-	resource.ParallelTest(t, *tc)
+	resource.Test(t, *tc)
```
Comment (cveticm) [inline internal/service/flexcluster/resource_test.go:25]: Parallelisation of tests has been removed as this causes errors where pluralDS expects 1 flex cluster to be present but an additional cluster is found from the other Acc test.

File: internal/service/flexcluster/plural_data_source.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+func (d *pluralDS) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var tfModel TFModelDSP
+	resp.Diagnostics.Append(req.Config.Get(ctx, &tfModel)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	connV2 := d.Client.AtlasV2
+	apiResp, _, err := connV2.FlexClustersApi.ListFlexClusters(ctx, tfModel.ProjectId.ValueString()).Execute()
```
Comment (AgustinBettati) [inline internal/service/flexcluster/plural_data_source.go]: q: Is this request paginated? If it is, would look into usage of `dsschema.AllPages` to see how we iterate over all results.

File: internal/service/flexcluster/resource_test.go [hunk 3]
```diff
@@ -136,11 +144,13 @@ func checksFlexCluster(projectID, clusterName string, terminationProtectionEnabl
...
-	checks = acc.AddAttrChecks(dataSourceName, checks, attrMap)
-	checks = acc.AddAttrSetChecks(resourceName, checks, attrSet...)
-	checks = acc.AddAttrSetChecks(dataSourceName, checks, attrSet...)
-	return resource.ComposeAggregateTestCheckFunc(checks...)
+	checks = acc.AddAttrChecks(dataSourcePluralName, checks, pluralMap)
+	for i := 0; i < flexClusterCount; i++ {
+		checks = acc.AddAttrSetChecks(resourceName, checks, "name", "termination_protection_enabled")
+		checks = acc.AddAttrSetChecks(dataSourceName, checks, "id")
+		checks = acc.AddAttrSetChecks(dataSourcePluralName, checks, "results.0.id")
+	}
```
Comment (AgustinBettati) [inline internal/service/flexcluster/resource_test.go]: Can these additional checks be handled as part of the `attrSet` that is passed to `CheckRSAndDS`?

File: internal/service/flexcluster/model.go [hunk 4]
```diff
@@ -41,6 +41,25 @@ func NewTFModel(ctx context.Context, apiResp *admin.FlexClusterDescription202411
...
 
+func NewTFModelDSP(ctx context.Context, projectID string, input *admin.PaginatedFlexClusters20241113) (*TFModelDSP, diag.Diagnostics) {
+	diags := &diag.Diagnostics{}
+	results := input.GetResults()
+	tfModels := make([]TFModel, len(results))
+	for i := range results {
+		item := &results[i]
+		tfModel, diagsLocal := NewTFModel(ctx, item)
+		diags.Append(diagsLocal...)
+		tfModels[i] = *tfModel
```
Comment (AgustinBettati) [inline internal/service/flexcluster/model.go]: would add an if to ensure tfModel is defined before appending to results, this avoids a potential  nil pointer dereference

File: internal/service/flexcluster/resource_test.go (hunk 1 above)
Comment (AgustinBettati) [reply internal/service/flexcluster/resource_test.go:25]: Would add as a comment so we have context in future reference

File: internal/service/flexcluster/resource_test.go [hunk 5]
```diff
@@ -115,16 +116,23 @@ func configBasic(projectID, clusterName, provider, region string, terminationPro
...
+func checksFlexCluster(projectID, clusterName string, terminationProtectionEnabled bool, flexClusterCount int) resource.TestCheckFunc {
 	checks := []resource.TestCheckFunc{checkExists()}
 	attrMap := map[string]string{
 		"project_id":                     projectID,
 		"name":                           clusterName,
 		"termination_protection_enabled": fmt.Sprintf("%v", terminationProtectionEnabled),
 	}
+	pluralMap := map[string]string{
+		"project_id": projectID,
+		"results.#":  "1",
```
Comment (AgustinBettati) [inline internal/service/flexcluster/resource_test.go:135]: is `flexClusterCount` param needed? I believe we are assuming config will only have 1 flex cluster

**Review by davidhou17 — APPROVED** [review]

File: internal/service/flexcluster/resource_test.go (hunk 1 above)
Comment (cveticm) [reply internal/service/flexcluster/resource_test.go:25]: Added comment

File: internal/service/flexcluster/plural_data_source.go (hunk 2 above)
Comment (cveticm) [reply internal/service/flexcluster/plural_data_source.go]: Good point, adjusted code 👍

File: internal/service/flexcluster/resource_test.go (hunk 3 above)
Comment (cveticm) [reply internal/service/flexcluster/resource_test.go]: Yes agreed, initial motivation was to iterate over and check all flex Clusters in pluralDS but I see now that this implementation does not even achieve that and testing only includes one flex cluster as is. Removed 👍

File: internal/service/flexcluster/model.go (hunk 4 above)
Comment (cveticm) [reply internal/service/flexcluster/model.go]: Added

File: internal/service/flexcluster/resource_test.go (hunk 5 above)
Comment (cveticm) [reply internal/service/flexcluster/resource_test.go:135]: Agreed, removed `flexClusterClount`

**Review by AgustinBettati — APPROVED** [review]
LGTM, thanks for the follow-ups

**Pull request merged**